internal/processing/istio/authorization_policy_processor.go

```diff
@@ -92,36 +100,59 @@ func generateAuthorizationPolicySpec(api *gatewayv1beta1.APIRule, rule gatewayv1
 		serviceName = *api.Spec.Service.Name
 	}
 
-	authorizationPolicySpec := builders.AuthorizationPolicySpecBuilder().
-		Selector(builders.SelectorBuilder().MatchLabels(processors.AuthorizationPolicyAppSelectorLabel, serviceName))
-
-	defaultScopeKeys := []string{"request.auth.claims[scp]", "request.auth.claims[scope]", "request.auth.claims[scopes]"}
-	for _, scope := range defaultScopeKeys {
-		generatedRule := generateAuthorizationPolicySpecRule(rule, scope, authorization)
-		authorizationPolicySpec.Rule(generatedRule)
-		// When scopes are defined, we need multiple rules for all of the above defaultScopeKeys
-		// When scopes are not defined, we just need one rule containing methods and paths
-		if generatedRule.Get().When == nil || len(generatedRule.Get().When) == 0 {
-			break
+	authorizationPolicySpecBuilder := builders.NewAuthorizationPolicySpecBuilder().
+		WithSelector(builders.NewSelectorBuilder().WithMatchLabels(processors.AuthorizationPolicyAppSelectorLabel, serviceName).Get())
+
+	// If RequiredScopes are configured, we need to generate a seperate Rule for each scopeKey in defaultScopeKeys
+	if len(authorization.RequiredScopes) > 0 {
+		for _, scopeKey := range defaultScopeKeys {
+			ruleBuilder := baseRuleBuilder(rule)
+			for _, scope := range authorization.RequiredScopes {
+				ruleBuilder.WithWhenCondition(
+					builders.NewConditionBuilder().WithKey(scopeKey).WithValues([]string{scope}).Get())
+			}
+
+			for _, aud := range authorization.Audiences {
+				ruleBuilder.WithWhenCondition(
+					builders.NewConditionBuilder().WithKey(AudienceKey).WithValues([]string{aud}).Get())
+			}
+
+			authorizationPolicySpecBuilder.WithRule(ruleBuilder.Get())
+		}
+	} else { // Only one AP rule should be generated for other scenarios
+		ruleBuilder := baseRuleBuilder(rule)
+		for _, aud := range authorization.Audiences {
+			ruleBuilder.WithWhenCondition(
+				builders.NewConditionBuilder().WithKey(AudienceKey).WithValues([]string{aud}).Get())
 		}
+		authorizationPolicySpecBuilder.WithRule(ruleBuilder.Get())
 	}
 
-	return authorizationPolicySpec.Get()
+	return authorizationPolicySpecBuilder.Get()
 }
 
-func generateAuthorizationPolicySpecRule(rule gatewayv1beta1.Rule, scope string, authorization *gatewayv1beta1.JwtAuthorization) *builders.Rule {
-	ruleBuilder := builders.RuleBuilder().RuleTo(builders.RuleToBuilder().
-		Operation(builders.OperationBuilder().Methods(rule.Methods).Path(rule.Path)))
+func withTo(b *builders.RuleBuilder, rule gatewayv1beta1.Rule) *builders.RuleBuilder {
```

Comment:
I think it would make sense to move the functions `withTo`, `withFrom` and `baseRuleBuilder` to the builders package. 
I think we can also think about splitting files like `istio.go` in the builders package in the future and even moving them to the scope where they are used, e.g. the istio processing package.

Reply:
I'm against placing those in builder package, as this are convenience funcs that make sense only here.
Splitting up is good idea